---
title: "Week6 assignment"
author: "Miriam Kretschmer"
date: "7 4 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(memes)
  library(Biostrings)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(PWMEnrich)
  library(ggplot2)
})
```


### Downloading ChIP peaks for Glucocorticoid receptor
```{r, eval=FALSE}
setwd("C:\\Users\\mkretschmer\\Documents\\Courses\\Bioinformatic_approaches\\Week6")
dir.create("peaks")
download.file("https://www.encodeproject.org/files/ENCFF676TDJ/@@download/ENCFF676TDJ.bed.gz", "GATA1_erythroblast_peak.bed.gz")
```

### Extract regions around peaks and obtain their sequence 
```{r}
peaks <- rtracklayer::import("GATA1_erythroblast_peak.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "Ensembl"
peaks_chr1 <- peaks[seqnames(peaks)=="1"]
peak_centers <- resize(peaks_chr1, fix="center", width=100)

ah <- AnnotationHub(localHub=TRUE)
genome <- ah[["AH68356"]]
peak_seqs <- memes::get_sequence(peak_centers, genome)
peak_seqs
```

## Motif instances in the GATA1 ChIP dataset
```{r}
#' findMotifInstances
#'
#' A wrapper around TFBSTools for scanning motif occurence, handling the 
#' coordinate conversion as `memes` does.
#' 
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with 
#'   coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```

```{r}
motifs <- query(MotifDb, "GATA1")

names(motifs)

motif <- motifs[["Mmusculus-JASPAR_CORE-Gata1-MA0035.2"]]

view_motifs(motif)
```

```{r}
moi <- findMotifInstances(peak_seqs, motif, mc.cores=2) # running with 2 threads
```
```{r}
length(peaks_chr1)
sum(overlapsAny(peaks_chr1, moi))

(100/length(peaks_chr1))*sum(overlapsAny(peaks_chr1, moi))
```
Out of 454 total peaks in Chromosome 1 of the GATA1 ChIP in erythroblasts, there are 424 peaks, so 93.39% that contain the GATA1 motif.


## Total motif occurrences in genome and their overlap with GATA1 ChIP peaks
```{r}
whole_genome <- readDNAStringSet("C:\\Users\\mkretschmer\\Documents\\Courses\\Bioinformatic_approaches\\Week3\\GRCm38v102_genome\\genome.fasta.gz")
genome_motifs <- findMotifInstances(seqs=whole_genome, motif=motif, mc.cores=4)
length(genome_motifs)

sum(overlapsAny(genome_motifs, peaks))

(100/length(genome_motifs))*sum(overlapsAny(genome_motifs, peaks))
```
Out of `r length(genome_motifs)` motif instances in the whole genome, `r sum(overlapsAny(genome_motifs, peaks))`, so 0.09 % overlap with peaks of the GATA1 ChIP dataset, 



```{r}
sessionInfo()
```